Fix embedded group

SebastienPeillet opened the pull request:
## Description

See #35872 

The embedded sub-group checked state is determined by the embedded project and not by the current project. So if the embedded project has an sub-group unchecked, there is no way to save it as checked in the current project.

As we handle layers visibility with the `embedded-invisible-layers` tag, I added a `embedded-unchecked-groups` to handle group checked state.

```xml
  <layer-tree-group>
    <customproperties/>
    <layer-tree-group name="group1" expanded="1" checked="Qt::Checked">
      <customproperties>
        <property key="embedded" value="1"/>
        <property key="embedded-invisible-layers"/>
        <property key="embedded-unchecked-groups" value="sub-group"/>
        <property key="embedded_project" value="./projet.qgs"/>
      </customproperties>
    </layer-tree-group>
```

I added elements in testEmbeddedGroup ([testqgslayertree.cpp](https://github.com/SebastienPeillet/QGIS/blob/9020f0dd97fcf2ef7c8e0f7ac5b5c0fd978a6067b/tests/src/core/testqgslayertree.cpp#L538)) to verify the sub-group checked state works.

Commit:
BUG: fix embedded sub_group state that are not saved in project

Commit:
Test on embedded subgroup visibility state

Commit:
Change QCOMPARE to QVERIFY

Commit:
REFAC: use embedded group tree index instead of group name to identify which group is unchecked

Commit:
REFAC: add id to QgsLayerTreeGroup, allow to embedded groups

DOC: changes on docstrings

Commit:
Sipify with new doc

3nids commented:
I added the needs documentation to test if the bot is working, do not consider any message

Commit:
sipify qgslayertreegroup

3nids commented on src/core/project/qgsproject.h:
```diff
@@ -653,7 +653,7 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      * \since QGIS 2.4
      */
-    QgsLayerTreeGroup *createEmbeddedGroup( const QString &groupName, const QString &projectFilePath, const QStringList &invisibleLayers, QgsProject::ReadFlags flags = QgsProject::ReadFlags() );
+    QgsLayerTreeGroup *createEmbeddedGroup( const QString &groupName, const QString &projectFilePath, const QStringList &invisibleLayers, const QStringList &uncheckedGroups, QgsProject::ReadFlags flags = QgsProject::ReadFlags() );
```
this an API break, no?

SebastienPeillet replied:
Yes, as I had to add the additionnal argument in each call (qgisapp.cpp, qgsproject.cpp)

3nids replied:
We should not break the API where it can be avoided.
You should add the additional argument at the end of the signature with a default value.

SebastienPeillet replied:
I fixed it !

3nids reviewed (commented):
How your solution handles groups having the same name either within the embed groups or between embed groups and the ones from the main project

SebastienPeillet commented:
> How your solution handles groups having the same name either within the embed groups or between embed groups and the ones from the main project

I forgot to put the reference of the first PR #35874. It has been closed by inactivity.

@nyalldawson already look at the code once, I point that problem. I added a id to groups to handle groups with the same name (see [comment](https://github.com/qgis/QGIS/pull/35874#discussion_r412233111))

3nids commented on src/core/layertree/qgslayertreegroup.h:
```diff
@@ -55,6 +55,21 @@ class CORE_EXPORT QgsLayerTreeGroup : public QgsLayerTreeNode
+    /**
+     * Generate the group's id, according to the groups's name (based on QgsMapLayer::generateId)
+     */
+    QString generateId( QString name ) const;
```
Does it make any sense to have this public?

SebastienPeillet replied:
Probably not, I moved the method to `protected`

3nids commented on src/core/layertree/qgslayertreegroup.h:
```diff
@@ -55,6 +55,21 @@ class CORE_EXPORT QgsLayerTreeGroup : public QgsLayerTreeNode
+    /**
+     * Sets the group's id.
+     */
+    void setId( const QString &id );
```
I don't think we should have this in the API. ID should be consistent in time.
I would rather add an optional argument to the constructor.

SebastienPeillet replied:
:+1: 

I removed the setId method as you suggest.

Now the constructor looks like : 
```h
// in .h file
QgsLayerTreeGroup( const QString &name = QString(), bool checked = true, const QString id = QString() );
```

and
```cpp
// in .cpp file
QgsLayerTreeGroup::QgsLayerTreeGroup( const QString &name, bool checked, const QString id )
  : QgsLayerTreeNode( NodeGroup, checked )
  , mName( name )
  , mId( ( id.isNull() ) ? generateId( name ) : id )
{
  connect( this, &QgsLayerTreeNode::visibilityChanged, this, &QgsLayerTreeGroup::nodeVisibilityChanged );
}
```

nyalldawson commented:
This is too dangerous for 3.18.0 - it needs to wait till post release

nyalldawson commented:
Just a heads-up @3nids -- this needs to be *very, very* carefully reviewed. The last change made to this code broke map theme support compatibility with older projects, which is why I've been extremely reluctant to review and merge this work.

3nids commented on src/core/layertree/qgslayertreeutils.h:
```diff
@@ -79,6 +79,9 @@ class CORE_EXPORT QgsLayerTreeUtils
     //! Gets invisible layers
     static QStringList invisibleLayerList( QgsLayerTreeNode *node );
 
+    //! Gets unchecked groups
```
```suggestion
    //! Returns the IDs of all unchecked groups
```

SebastienPeillet replied:
:+1:

3nids commented on src/core/layertree/qgslayertreeutils.cpp:
```diff
@@ -382,6 +382,28 @@ QStringList QgsLayerTreeUtils::invisibleLayerList( QgsLayerTreeNode *node )
+    const auto constChildren = QgsLayerTree::toGroup( node )->children();
+    for ( QgsLayerTreeNode *child : constChildren )
+    {
+      if ( QgsLayerTree::isGroup( child ) )
```
I think the method should be called recursively here, so that we can have nested groups which current implementation does not consider if I am correct.

SebastienPeillet replied:
Actually it does ! The method is called recursively on [line 400](https://github.com/qgis/QGIS/pull/41582/files#diff-1ef9764fa0bcd5fabfbf51974955fec2940df478ac3c7f27bacd70bb6d194b0eR400) with each group node.

Commit:
Fix after feedback, change createEmbeddedGroup with additional default parameter

remove setId method, id will be added as an optionnal argument in the constructor

overlooked mistake in doc

id shouldn't be translated

Commit:
Try to handle old project version without break

SebastienPeillet commented:
As feared by @nyalldawson, there is uncompatibility with old project version when fixing theme functionnalities.

I add the [last commit](https://github.com/qgis/QGIS/pull/41582/commits/69130156b8149dca9bb28016fa57e0528d9bd139) to handle old project with the same functionalities (and the same unabilities).

To handle theme with the new group id, I changed qgsmapthemecollection.cpp : 

```cpp
static QString _groupId( QgsLayerTreeNode *node )
{
  QStringList lst;
  while ( node->parent() )
  {
    lst.prepend( node->name() );
    lst.prepend( QgsLayerTree::toGroup( node )->id() ); // instead of lst.prepend( node->name() );
    node = node->parent();
  }
  return lst.join( '/' );
}
```
And to support old project, I changed the [readXml](https://github.com/qgis/QGIS/pull/41582/files#diff-e3dcfa73de11b3fb7664612b93f0590469d034d2584d4049dd487a98bb02c6c6R290) method on qgslayertreegroup : 

```
QString id = ( element.attribute( QStringLiteral( "id" ) ).isNull() ) ? name : element.attribute( QStringLiteral( "id" ) );
// ...
QgsLayerTreeGroup *groupNode = new QgsLayerTreeGroup( name, checked, id );
```
When it's an old project, there will be no `id` tag in the group, so we use the name group as id to keep old functionalities (theme are using group name and the id will be equivalent). When the project is saved, the id will be saved and not updated. So there is no update on a old project. It stay functional as before.

With a new project the id will be a generated one to benefit to the fix

Commit:
missing QUuid include
